add release 132 blog post (#156)
conan-io/conan-io.github.io: _posts/2020-12-14-New-conan-release-1-32.markdown

---
layout: post 
comments: false 

> title: "New validate() Method, First Configurable Generator (msbuild)

Reviewer: Add: "Conan 1.32" release in the title

(Experimental), Renamed Multiple Toolchains & Generators (Methods & Classes), 2
New Meson Toolchains, Improve Lockfile Support for Private Requirements, Support
for build_requires to Affect package_id (Workaround)"
---

Conan 1.32.0 features a new strategy for defining, detecting, and handling
invalid configurations. This includes a new method named `validate()`, and a new
`package_id` mode named `BINARY_INVALID`.  It also includes a big refactor of

> our future toolchain and generators strategy. It also includes a new capability,

Reviewer 2: repetition of "it also includes"

and 2 new generators for Meson. We've also added some subcommands for dealing
with recipes and packages which have no remote associated, and made it possible
to deal with lockfiles which contain the same package multiple times in the
graph (possible via private requirements).  Finally, we're excited to show a
proven workaround which satisfies a long-standing and high-priority request,
which is to optionally include `build_requires` into the `package_id`.  There

> are some important side-effects and caveats, to be aware of, so read on to learn

Reviewer 2: comma needed?
```suggestion
are some important side-effects and caveats to be aware of, so read on to learn
```

Author: I don't think so, but could be correct. In school, the lesson I remember hearing the most was that commas are highly over-used, and every time you put a comma because you pause there in your mind, you should stop and consider if it's actually necessary grammatically.  It's usually not. In fact, the existing comma might even be superfluous as well.

about the details.

## New Strategy for Invalid Configurations

In recent months, there has been growing discussion in the community about some
awkward challenges surrounding the definition, detection, and handling invalid
or impossible configurations. Historically, the general recommendation has been
to handle these cases by raising `ConanInvalidConfiguration` errors during the
`configure()` method, but in practice there are some checks which simply cannot
be run there because of the state of the dependency graph. Also, with the

> `compatible_packages` feature, there there are cases where a recipe might not be

Reviewer 2: repeated "there", and "when" instead of "when"? (not sure)
```suggestion
`compatible_packages` feature, there are cases when a recipe might not be
```

Author: "when" is better here, good catch.  "where" and "when"  are often used interchangeably (and sub-optimally) after flexible words like "cases" and "situations".  If the previous word had been "times", then obviously "times when" would be more correct, but people still very often say "there are times where".

> `compatible_packages` feature, there there are cases where a recipe might not be

Reviewer 3: ```suggestion
`compatible_packages` feature, there are cases where a recipe might not be
```

able to build, but might be configured to provide a compatible package instead.

The existing workarounds for these cases were unfortunate, so we've now added
some first-class-features around them. The first is to add a new method to

> Conanfile named `validate()` method which executes after the graph has been

Reviewer: A snippet of code showing the ``validate()`` method in action, and the output of the ``conan info`` and ``conan create`` command would be valuable.

Reviewer 2: yeah, that would be fine or maybe  just a direct link to the docs section

fully evaluated solving the previous problems. The second is to add a new
`package_id` mode named `BINARY_INVALID`.  Moving forward, the `configure()`
method should still be used for actually setting configurations, but the
`validate()` method should take over all responsibility for raising
`ConanInvalidConfiguration` errors. In these cases, the `BINARY_INVALID` is now
reported as the state in the output of  `conan info` command, as opposed the
previous behavior which was for this command to fail with return code 6 which

> was to helpful at all. This is a major improvement as some automation cases

Reviewer 2: "was not helpfult at all" maybe? I didn't understand this sentence

> (such as Conan Center) use `conan info` to make decisions about what to build,

Reviewer 2: ```suggestion
(such as ConanCenter) use `conan info` to make decisions about what to build,
```

and parsing json output is more desirable than interpreting return codes.

## Refactoring Toolchains and Generators

It's important to highlight that a lot of experimental classes were moved around
into different namespaces in this release. Please refer to the changelog and
docs for all the specifics, but here's a simple summary.

> We've created four new top level folders under the namespace "conan.tools":

Reviewer: Highlight that it is "conan" and not "conans", and say this is a step towards Conan 2.0,

Author: added a whole paragraph about it.

* cmake
* gnu
* meson
* microsoft

We'll be adding to this structure in the future, and the decision was that
organizing these tools for integrations by "vendor" was a lot more intuitive
than trying to organize them by "purpose" or "functionality". Such strategies
have proven to be awkward in the past because so many tools and utility
functions are multi-purpose.

We've started using these directories by putting all the respective toolchains
classes into in them.  We've also begun moving forward on a new naming
convention for generators, starting with the experimental `msbuild` generator.

> It has been renamed to `MsBuildDeps`, and we'll talk more about why we did this

Reviewer: ```suggestion
It has been renamed to `MSBuildDeps`, and we'll talk more about why we did this
```

shortly. We've also created a new generator named `CMakeDeps` in the `cmake`
directory, however, it's really just an alias which redirects to the
`cmake_find_package_multi` generator. We've already been promoting that as the
top recommendation for most use-cases of `CMake` generators, and we'll now be
promoting that people use it from this new import and new name (`CMakeDeps`).

> Also, we've added a new toolchain called `MesonToolchain`, which includes

Reviewer 3: ```suggestion
Also, we've added a new toolchain called `MesonToolchain`, which
```

Author: Good catch, in-fact, caused me to go back and restructure (and correct) everything about meson.

worse than one of the existing generators which has no configuration at all.

## Toolchain-Related Methods Renamed

The revelation process described above also allowed us to step back and
recognize a simplification that could be made.  While Toolchains and Generators
are definitely distinct categories of information from the perspective of the
build, we've decided that they're not substantially different in terms of their
essence in Conan. As such, we've decided to refactor the implementations around
them to treat them in a similar way.  This dramatically reduces internal
complexity and increases consistency in a several ways, but simply put, it's
less to learn.

The overall idea here is that the process for both things is still the
same:

* Analyze the configuration and dependency data Conan has gathered
* Generate files
  
They're just two sub categories of the same thing. So, at that point, it seemed
fairly obvious to rename the still-experimental `def toolchain()` method in
`conanfile.py` to `def generate()`. For the same reasons, we decided that the
"primary interface method" on both Toolchains and Generators moving forward

> should just be `generate()` as well. It's simple, consistent, and descriptive.

Reviewer: Comment that the ``write_toolchain_files()`` has also been renamed to ``generate()``.

Reviewer: Show a full example of the generate() recipe method, instantiating the ``MSBuildDeps``, for example, customizing it with "ReleaseShared", and calling ``generate()``. I think it will be more illustrative than the text. Show the new ``MSBuild()`` build helper, importing from the new space and highlighting it is a new, different one (can be a comment in the code)

Author: There's so many new ideas around generators now, it feels like something that should be part of a new/separate blog post showing the new vision of `MSBuildToolchain`, `MSBuildDeps`, and `MSBuild` helper together, with a complete working example (Maybe a working example in the examples repository? )  I want to fully explain the intended benefits of the new configurability, talk about how we're sidestepping the old problem of per-configuration `import` statements, and talk about the implications for multi-project solutions (like the ability to generate a separate props-per-project with partial deps).  Also, I need to take time exercising the feature myself to create such an example and write about it.  WDYT?

Author: I've added a small example just showing how the toolchain and generator calls will look nicely side-by-side now.  Again, let me know what you think.

Reviewer: Yes, agree. At the moment just add something like:
```python
from conans import ConanFile
from conan.tools.microsoft import MSBuildDeps, MSBuildToolchain

class Pkg(ConanFile):
    requires = "boost/1.72.0", "poco/1.9.4"
    # The generate method executes at "conan install"
    def generate(self):
        # generates conan_boost.props, conan_poco.props for dependencies
        deps= MSBuildDeps(self)
        deps.generate()
        # generates conantoolchain.props from current settings
        tc= MSBuildToolchain(self)
        tc.generate()
```

I think this is very simple, but also captures visually many interesting things from the changes:
- The new imports
- The controlled and explicit generation of both deps information and toolchain information

I am trying to make the blog post a bit more visually light-weight, a couple of code snippets like this can really help to give some context to readers. Show me the code! :)

Author: > a couple of code snippets like this can really help to give some context to readers. Show me the code! :)

agree

>   The generate method executes at "conan install"  

I think this comment is ok, but I think this might be confusing to some people. Maybe people will think they have to run `conan install` manually for the `generate()` method to work.  I don't think it's obvious to everyone that there is an implicit equivalent of `conan install` which happens when running `conan create`, and whenever something talks about `conan install` command, it's also including this implicit install of `conan create`.  I think most people will understand that, but I don't think it's "abundantly clear". 

>  generates conan_boost.props, conan_poco.props for dependencies  

We should also mention the `conan_deps.props` here which includes "all other props" right?

Reviewer: > I think this comment is ok, but I think this might be confusing to some people. Maybe people will think they have to run conan install manually for the generate() method to work. I don't think it's obvious to everyone that there is an implicit equivalent of conan install which happens when running conan create, and whenever something talks about conan install command, it's also including this implicit install of conan create. I think most people will understand that, but I don't think it's "abundantly clear".

Yes, could be, you can leave it out, or try to clarify with a couple of words

> We should also mention the conan_deps.props here which includes "all other props" right?

Sure, why not, not super critical, but nice to have it.

> ## Including build_requires and package_id (Experimental)

Reviewer 2: "in" package id?

> There has been an outstanding feature request for a LONG time which has seemed

Reviewer 2: ```suggestion
There has been an outstanding feature request for a **long** time which has seemed
```

are definitely distinct categories of information from the perspective of the
build, we've decided that they're not substantially different in terms of their
essence in Conan. As such, we've decided to refactor the implementations around
them to treat them in a similar way.  This dramatically reduces internal
complexity and increases consistency in a several ways, but simply put, it's
less to learn.

The overall idea here is that the process for both things is still the
same:

* Analyze the configuration and dependency data Conan has gathered
* Generate files
  
They're just two sub categories of the same thing. So, at that point, it seemed
fairly obvious to rename the still-experimental `def toolchain()` method in
`conanfile.py` to `def generate()`. For the same reasons, we decided that the
"primary interface method" to trigger the generation of files on both Toolchains
and Generators moving forward should just be `generate()` as well. So, for
toolchains, the old method `write_toolchain_files()` has been renamed to
`generate()`.  It's simple, consistent, and descriptive. Here's an example of
using the two together:

> 
>     def generate(self):
>       MSBuildDeps(self).generate()
>       MSBuildToolchain(self).generate()
> 
> Furthermore, going back to the topic of deprecating of the lower-case `name`

Reviewer: ```suggestion

    from conans import ConanFile
    from conan.tools.microsoft import MSBuildDeps, MSBuildToolchain

    class Pkg(ConanFile):
        requires = "boost/1.72.0", "poco/1.9.4"
        # The generate method creates files that the build can use
        def generate(self):
            # generates conan_boost.props, conan_poco.props for dependencies
            # and conandeps.props that aggregates them
            deps= MSBuildDeps(self)
            deps.generate()
            # generates conantoolchain.props from current settings
            tc= MSBuildToolchain(self)
            tc.generate()

Furthermore, going back to the topic of deprecating of the lower-case `name`
```

---

> layout: post 
> comments: false 
> title: "Conan 1.32: New validate() Method, First Configurable Generator (msbuild)
> (Experimental), Renamed Multiple Toolchains & Generators (Methods & Classes), 2
> New Meson Classes, Improve Lockfile Support for Private Requirements, Support
> for build_requires to Affect package_id (Workaround)"
> ---

Reviewer: ```suggestion
layout: post 
comments: false 
title: "Conan 1.32: New validate() Method, First configurable Generator (MSBuildDeps), 
Renamed Multiple Toolchains & Generators, 2 New Meson Classes, Improve Lockfile
Support for Private Requirements, Support for build_requires to Affect package_id (Workaround)"
---
```

build, we've decided that they're not substantially different in terms of their
essence in Conan. As such, we've decided to refactor the implementations around
them to treat them in a similar way.  This dramatically reduces internal
complexity and increases consistency in a several ways, but simply put, it's
less to learn.

The overall idea here is that the process for both things is still the
same:

* Analyze the configuration and dependency data Conan has gathered
* Generate files
  
They're just two sub categories of the same thing. So, at that point, it seemed
fairly obvious to rename the still-experimental `def toolchain()` method in
`conanfile.py` to `def generate()`. For the same reasons, we decided that the
"primary interface method" to trigger the generation of files on both Toolchains
and Generators moving forward should just be `generate()` as well. So, for
toolchains, the old method `write_toolchain_files()` has been renamed to
`generate()`.  It's simple, consistent, and descriptive. Here's an example of
using the two together:

    from conans import ConanFile
    from conan.tools.microsoft import MSBuildDeps, MSBuildToolchain

> 
>     class Pkg(ConanFile):
>         requires = "boost/1.72.0", "poco/1.9.4"
>         # The generate method creates files that the build can use
>         def generate(self):
>             # generates conan_boost.props, conan_poco.props for dependencies

Reviewer: ```suggestion

    class Pkg(ConanFile):
        settings = "os", "compiler", "build_type", "arch"
        requires = "boost/1.72.0", "poco/1.9.4"
        
        # The generate method creates files that the build can use
        def generate(self):
            # generates conan_boost.props, conan_poco.props for dependencies
```
